File: Degree/Semester-4/OOP/5_prog/LibraryCatalogManager.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class LibraryCatalogManager {
    static class Book {
        final String isbn;
        final String title;
        final String author;
        boolean available;

        Book(String isbn, String title, String author) {
            this.isbn = isbn;
            this.title = title;
            this.author = author;
            this.available = true;
        }

        @Override
        public String toString() {
            return String.format("%s | %s by %s | %s", isbn, title, author, available ? "Available" : "Issued");
        }
    }

    static class Member {
        final int memberId;
        final String name;
        final List<String> borrowedIsbns = new ArrayList<>();

        Member(int memberId, String name) {
            this.memberId = memberId;
            this.name = name;
        }

        @Override
        public String toString() {
            return String.format("#%d %s | borrowed: %s", memberId, name, borrowedIsbns);
        }
    }

    static class LibraryService {
        private final Map<String, Book> booksByIsbn = new HashMap<>();
        private final Map<Integer, Member> membersById = new HashMap<>();

        void registerBook(Book book) {
            booksByIsbn.put(book.isbn, book);
        }

        void registerMember(Member member) {
            membersById.put(member.memberId, member);
        }

        boolean issueBook(int memberId, String isbn) {
            Member member = membersById.get(memberId);
            Book book = booksByIsbn.get(isbn);

            if (member == null || book == null || !book.available) {
                return false;
            }

            book.available = false;
            member.borrowedIsbns.add(isbn);
            return true;
        }

        boolean returnBook(int memberId, String isbn) {
            Member member = membersById.get(memberId);
            Book book = booksByIsbn.get(isbn);

            if (member == null || book == null || !member.borrowedIsbns.contains(isbn)) {
                return false;
            }

            member.borrowedIsbns.remove(isbn);
            book.available = true;
            return true;
        }

        List<Book> searchByKeyword(String keyword) {
            String key = keyword.toLowerCase();
            List<Book> result = new ArrayList<>();

            for (Book book : booksByIsbn.values()) {
                if (book.title.toLowerCase().contains(key) || book.author.toLowerCase().contains(key)) {
                    result.add(book);
                }
            }
            return result;
        }

        void printState() {
            System.out.println("Books:");
            booksByIsbn.values().forEach(book -> System.out.println("  " + book));

            System.out.println("Members:");
            membersById.values().forEach(member -> System.out.println("  " + member));
        }
    }

    public static void main(String[] args) {
        LibraryService library = new LibraryService();

        library.registerBook(new Book("978-0134685991", "Effective Java", "Joshua Bloch"));
        library.registerBook(new Book("978-1617294945", "Spring in Action", "Craig Walls"));
        library.registerBook(new Book("978-1492056270", "Designing Data-Intensive Applications", "Martin Kleppmann"));

        library.registerMember(new Member(1, "Aarav"));
        library.registerMember(new Member(2, "Meera"));

        System.out.println("=== Library Catalog Manager ===");
        System.out.println("Issue Effective Java to member #1: " + library.issueBook(1, "978-0134685991"));
        System.out.println("Issue Effective Java to member #2 (should fail): " + library.issueBook(2, "978-0134685991"));
        System.out.println("Return Effective Java from member #1: " + library.returnBook(1, "978-0134685991"));
        System.out.println("Issue Effective Java to member #2 (should pass): " + library.issueBook(2, "978-0134685991"));

        System.out.println("\nSearch keyword 'data':");
        List<Book> matches = library.searchByKeyword("data");
        if (matches.isEmpty()) {
            System.out.println("  No books found.");
        } else {
            matches.forEach(book -> System.out.println("  " + book));
        }

        System.out.println();
        library.printState();
    }
}

File: Degree/Semester-4/OOP/Project/5_prog/LibraryCatalogManager.cpp
#include <iostream>
#include <map>
#include <vector>
#include <string>
#include <algorithm>

using namespace std;

struct Book {
    string isbn;
    string title;
    string author;
    bool available;

    Book(const string& isbn, const string& title, const string& author)
        : isbn(isbn), title(title), author(author), available(true) {}

    string toString() const {
        return isbn + " | " + title + " by " + author + " | " + (available ? "Available" : "Issued");
    }
};

struct Member {
    int memberId;
    string name;
    vector<string> borrowedIsbns;

    Member(int memberId, const string& name)
        : memberId(memberId), name(name) {}

    string toString() const {
        string borrowed = "";
        for (size_t i = 0; i < borrowedIsbns.size(); i++) {
            if (i > 0) borrowed += ", ";
            borrowed += borrowedIsbns[i];
        }
        return "#" + to_string(memberId) + " " + name + " | borrowed: [" + borrowed + "]";
    }
};

class LibraryService {
private:
    map<string, Book> booksByIsbn;
    map<int, Member> membersById;

public:
    void registerBook(const Book& book) {
        booksByIsbn.insert({book.isbn, book});
    }

    void registerMember(const Member& member) {
        membersById.insert({member.memberId, member});
    }

    bool issueBook(int memberId, const string& isbn) {
        if (membersById.find(memberId) == membersById.end() ||
            booksByIsbn.find(isbn) == booksByIsbn.end() ||
            !booksByIsbn.at(isbn).available) {
            return false;
        }

        booksByIsbn.at(isbn).available = false;
        membersById.at(memberId).borrowedIsbns.push_back(isbn);
        return true;
    }

    bool returnBook(int memberId, const string& isbn) {
        if (membersById.find(memberId) == membersById.end() ||
            booksByIsbn.find(isbn) == booksByIsbn.end()) {
            return false;
        }

        auto& borrowed = membersById.at(memberId).borrowedIsbns;
        auto it = find(borrowed.begin(), borrowed.end(), isbn);
        if (it == borrowed.end()) {
            return false;
        }

        borrowed.erase(it);
        booksByIsbn.at(isbn).available = true;
        return true;
    }

    vector<Book> searchByKeyword(const string& keyword) const {
        string key = keyword;
        transform(key.begin(), key.end(), key.begin(), ::tolower);
        vector<Book> result;

        for (const auto& pair : booksByIsbn) {
            string title = pair.second.title;
            string author = pair.second.author;
            transform(title.begin(), title.end(), title.begin(), ::tolower);
            transform(author.begin(), author.end(), author.begin(), ::tolower);

            if (title.find(key) != string::npos || author.find(key) != string::npos) {
                result.push_back(pair.second);
            }
        }
        return result;
    }

    void printState() const {
        cout << "Books:" << endl;
        for (const auto& pair : booksByIsbn) {
            cout << "  " << pair.second.toString() << endl;
        }

        cout << "Members:" << endl;
        for (const auto& pair : membersById) {
            cout << "  " << pair.second.toString() << endl;
        }
    }
};

int main() {
    LibraryService library;

    library.registerBook(Book("978-0134685991", "Effective Java", "Joshua Bloch"));
    library.registerBook(Book("978-1617294945", "Spring in Action", "Craig Walls"));
    library.registerBook(Book("978-1492056270", "Designing Data-Intensive Applications", "Martin Kleppmann"));

    library.registerMember(Member(1, "Aarav"));
    library.registerMember(Member(2, "Meera"));

    cout << "=== Library Catalog Manager ===" << endl;
    cout << "Issue Effective Java to member #1: " << (library.issueBook(1, "978-0134685991") ? "true" : "false") << endl;
    cout << "Issue Effective Java to member #2 (should fail): " << (library.issueBook(2, "978-0134685991") ? "true" : "false") << endl;
    cout << "Return Effective Java from member #1: " << (library.returnBook(1, "978-0134685991") ? "true" : "false") << endl;
    cout << "Issue Effective Java to member #2 (should pass): " << (library.issueBook(2, "978-0134685991") ? "true" : "false") << endl;

    cout << "\nSearch keyword 'data':" << endl;
    vector<Book> matches = library.searchByKeyword("data");
    if (matches.empty()) {
        cout << "  No books found." << endl;
    } else {
        for (const auto& book : matches) {
            cout << "  " << book.toString() << endl;
        }
    }

    cout << endl;
    library.printState();

    return 0;
}
